core/payment-driver/erc20/src/driver/cron.rs

```diff
@@ -0,0 +1,163 @@
+                    "ZkSync transaction verification failed. tx_details={:?} error={}",
+                    tx,
+                    err
+                );
+                dao.transaction_failed(&tx.tx_id).await;
+                for order_id in order_ids.iter() {
+                    dao.payment_failed(order_id).await;
+                }
+                return;
+            }
+
+            // TODO: Add token support
+            let platform = network::network_token_to_platform(Some(network), None).unwrap(); // TODO: Catch error?
+            let details = match wallet::verify_tx(&tx_hash, network).await {
+                Ok(a) => a,
+                Err(e) => {
+                    log::warn!("Failed to get transaction details from erc20, creating bespoke details. Error={}", e);
+
+                    let first_payment: PaymentEntity = match dao.get_first_payment(&tx_hash).await {
+                        Some(p) => p,
+                        None => continue,
+                    };
+
+                    //Create bespoke payment details:
+                    // - Sender + receiver are the same
+                    // - Date is always now
+                    // - Amount needs to be updated to total of all PaymentEntity's
+                    let mut details = utils::db_to_payment_details(&first_payment);
+                    details.amount = payments
+                        .into_iter()
+                        .map(|payment| utils::db_amount_to_big_dec(payment.amount.clone()))
+                        .sum::<BigDecimal>();
+                    details
+                }
+            };
+            let tx_hash = hex::decode(&tx_hash[2..]).unwrap();
+
+            if let Err(e) = bus::notify_payment(name, &platform, order_ids, &details, tx_hash).await
+            {
+                log::error!("{}", e)
+            };
+        }
+    }
+}
+
+pub async fn process_payments_for_account(dao: &Erc20Dao, node_id: &str, network: Network) {
+    log::trace!(
+        "Processing payments for node_id={}, network={}",
+        node_id,
+        network
+    );
+    let payments: Vec<PaymentEntity> = dao.get_pending_payments(node_id, network).await;
+    let mut nonce = U256::from(0);
+    if !payments.is_empty() {
+        log::info!(
+            "Processing payments. count={}, network={} node_id={}",
+            payments.len(),
+            network,
+            node_id
+        );
+        let network_nonce =
+            wallet::get_network_nonce(crate::erc20::utils::str_to_addr(&node_id).unwrap(), network)
+                .await
+                .unwrap();
+        let db_nonce = dao.get_next_nonce(&node_id, network).await.unwrap();
+        nonce = if network_nonce > db_nonce {
+            network_nonce
+        } else {
+            db_nonce
+        };
+        log::debug!("Payments: nonce={}, details={:?}", &nonce, payments);
+    }
+    for payment in payments {
+        handle_payment(&dao, payment, &mut nonce).await;
+    }
```

Comment:
Pulling this loop into the `if` above would allow to remove the ugly way of setting nonce (0 first, then re-assigning).

Reply:
i'm not sure about this:
- without if it will query network and database each time
- with if you still need the ugly nonce = 0

Reply:
I think you misunderstood my suggestion. Moving this loop into the `if` doesn't change anything about the function behavior because the loop wouldn't run either way if `payments` is empty. And no, you don't need the ugly `nonce = 0`. Just take a look [here](https://github.com/golemfactory/yagna/compare/mwu/new-erc20...wiezzel/put_for_inside_if_in_process_payments_for_account_in_erc20_driver?expand=1).

Reply:
oops, my bad.. i read it the other way around /shy

Great idea! updated :)